File: 02-UML-Class-Diagrams/src/java/chapter02/Book.java
package chapter02;

/**
 * UML: Composition target + Association source.
 * Library *-- Book  (Library owns Book — composition)
 * Book --> Author   (Book knows Author — association)
 */
public class Book {
    private String title;
    private String isbn;
    private Author author;  // Association: Book knows its Author

    public Book(String title, String isbn, Author author) {
        this.title = title;
        this.isbn = isbn;
        this.author = author;
    }

    public String getTitle() { return title; }
    public String getIsbn() { return isbn; }
    public Author getAuthor() { return author; }

    public void displayInfo() {
        System.out.println("[Book] " + title + " by " + author.getName() + " (ISBN: " + isbn + ")");
    }

    @Override
    public String toString() {
        return title + " by " + author.getName();
    }
}

File: 02-UML-Class-Diagrams/src/go/book.go
package main

import "fmt"

// Composition target: Library owns Books.
// Association source: Book knows its Author (pointer = doesn't own).

type Book struct {
	Title  string
	ISBN   string
	Author *Author // Association: knows author, doesn't own
}

func NewBook(title, isbn string, author *Author) *Book {
	return &Book{Title: title, ISBN: isbn, Author: author}
}

func (b *Book) DisplayInfo() {
	fmt.Printf("[Book] %s by %s (ISBN: %s)\n", b.Title, b.Author.Name, b.ISBN)
}
